source/fuzz/fuzzer_pass_add_copy_memory_instructions.cpp

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2020 Vasyl Teliman
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/fuzzer_context.h"
+#include "source/fuzz/fuzzer_pass_add_copy_memory_instructions.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "source/fuzz/transformation_add_copy_memory.h"
+#include "source/fuzz/transformation_add_local_variable.h"
+
+namespace spvtools {
+namespace fuzz {
+
+FuzzerPassAddCopyMemoryInstructions::FuzzerPassAddCopyMemoryInstructions(
+    opt::IRContext* ir_context, TransformationContext* transformation_context,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}
+
+FuzzerPassAddCopyMemoryInstructions::~FuzzerPassAddCopyMemoryInstructions() =
+    default;
+
+void FuzzerPassAddCopyMemoryInstructions::Apply() {
+  ForEachInstructionWithInstructionDescriptor(
+      [this](opt::Function* function, opt::BasicBlock* /*unused*/,
+             opt::BasicBlock::iterator instr_it,
+             const protobufs::InstructionDescriptor& /*unused*/) {
+        const auto& instr = *instr_it;
+
+        // Omit the instruction if it doesn't have either the result id or
+        // the type id.
+        if (!instr.result_id() || !instr.type_id()) {
+          return;
+        }
+
+        const auto* type_instr = GetIRContext()->get_def_use_mgr()->GetDef(instr.type_id());
+        if (!type_instr || type_instr->IsOpaqueType() || type_instr->opcode() != SpvOpTypePointer) {
```

Comment:
Since instr.type_id was not zero, type_instr cannot be 0.  I would thus prefer you to assert that before the if, rather than have it as an if condition.

Reply:
Fixed in 2938026.